server/raft.go

```diff
@@ -3383,35 +3383,37 @@ func (n *raft) processAppendEntry(ae *appendEntry, sub *subscription) {
 		n.updateLeadChange(false)
 	}
 
-	if (isNew && ae.pterm != n.pterm) || ae.pindex != n.pindex {
-		// Check if this is a lower index than what we were expecting.
-		if ae.pindex < n.pindex {
+	if ae.pterm != n.pterm || ae.pindex != n.pindex {
+		// Check if this is a lower or equal index than what we were expecting.
+		if ae.pindex <= n.pindex {
 			n.debug("AppendEntry detected pindex less than ours: %d:%d vs %d:%d", ae.pterm, ae.pindex, n.pterm, n.pindex)
 			var ar *appendEntryResponse
-
-			// An AppendEntry is stored at seq=ae.pindex+1. This can be checked when eae != nil, eae.pindex==ae.pindex.
-			seq := ae.pindex + 1
 			var success bool
-			if eae, _ := n.loadEntry(seq); eae == nil {
+
+			if n.commit > 0 && ae.pindex <= n.commit {
+				// If we have already committed this entry, just mark success.
+				success = true
+			} else if eae, _ := n.loadEntry(ae.pindex); eae == nil {
 				// If terms are equal, and we are not catching up, we have simply already processed this message.
 				// So we will ACK back to the leader. This can happen on server restarts based on timings of snapshots.
 				if ae.pterm == n.pterm && !catchingUp {
 					success = true
+				} else if ae.pindex == n.pindex {
```

Comment:
I don't get this bit. 
If our previous index already matches, what are we truncating?

Reply:
Realistically, only updating the `pterm` when the passed `index == n.pindex`.

We could possibly also just set `n.pterm = ae.pterm` directly instead, but not sure why it was previously chosen to be a call to `truncateWAL` (so am just doing the same thing as what was done previously).